File: from-0801-to-0900/0872-leaf-similar-trees/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        def get_leaves(root) -> List[int]:
            leaves = []
            def dfs(node: TreeNode) -> None:
                nonlocal leaves
                if node.left is None and node.right is None:
                    leaves.append(node.val)
                    return
                if node.left:
                    dfs(node.left)
                if node.right:
                    dfs(node.right)

                return

            dfs(root)
            return leaves

        return get_leaves(root1) == get_leaves(root2)
        

File: from-0801-to-0900/0872-leaf-similar-trees/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
    var getLeafVals func(node *TreeNode, leafVals *[]int)
    getLeafVals = func(node *TreeNode, leafVals *[]int) {
        if node == nil {
            return
        }
        if node.Left == nil && node.Right == nil {
            *leafVals = append(*leafVals, node.Val)
            return
        } 
        getLeafVals(node.Left, leafVals)
        getLeafVals(node.Right, leafVals)
    }

    leafVals1, leafVals2 := []int{}, []int{}
    getLeafVals(root1, &leafVals1)
    getLeafVals(root2, &leafVals2)

    return reflect.DeepEqual(leafVals1, leafVals2)
}